### PR TITLE
perf(*) reduce to minimal string concat per request

### PR DESCRIPTION
**Pull request by fffonion**
QPS on packet.net baremetal 4c8t E3-1578L v5 @ 2.00GHz
vanilla kong: 44917.29
master: 7586.99
this branch: ~30018.47~ 37244.20 (after each worker is warmed up with cache)

- move key search to a table based search, expensive string operations now only occur once per label group per worker. this is basically trading cpu with memory and requires measurement of memory consumption on high cardinality.
- lua-resty-counter is used by default
- reduced function call in hot path

### CODE REVIEW

**Comment by fffonion** [comment]
Screen Shot 2020-03-12 at 8 18 42 PM

exporter.lua now appears only less than 10% in stacktraces

**Comment by fffonion** [comment]
Screen Shot 2020-03-12 at 8 28 42 PM

exporter.lua now appears only less than 10% in stacktraces
[1.svg.gz](https://github.com/Kong/kong-plugin-prometheus/files/4324137/1.svg.gz)

**Comment by fffonion** [comment]
scrape lock analysis:

without scrape:
```
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   269.84us  156.67us   3.16ms   91.81%
    Req/Sec    19.36k     2.04k   22.84k    75.12%
  387239 requests in 10.10s, 69.78MB read
Requests/sec:  38342.68
Transfer/sec:      6.91MB
```

with scrape every one second
```
Running 10s test @ http://10.11.9.129:8000
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     0.93ms    7.12ms 134.74ms   98.91%
    Req/Sec    18.99k     2.61k   23.10k    79.21%
  381683 requests in 10.10s, 68.78MB read
Requests/sec:  37790.38
Transfer/sec:      6.81MB
```

there's *some* impact especially with p95 latencies, but the overall QPS doesn't drop a lot. the scrape acquires the shdict lock several times, but on hot path counter/gauge increments are lock free.

**Comment by hbagdi** [comment]
@fffonion Did you do any measurement on the amount of memory consumption or how it will increase as more routes/services are added?

**Comment by fffonion** [comment]
Yeah. A three-label counter takes around 250 ~ 400 bytes for a single worker.

For example, in a 4 worker cluster:
```
status{"label1"="value1","label2"="value2","label3"="value3"} 1
status{"label1"="value4","label2"="value5","label3"="value6"} 2
```
will take 3200 bytes in total.

```
status{"label1"="value1","label2"="value2","label3"="value3"} 1
status{"label1"="value1","label2"="value2","label3"="value4"} 2
```
This will takes less if labels diverges at last. But the scale is still like ~200 ish per metrics. The majority of space is pre-formatted string. We store the full `status{"label1"="value1","label2"="value2","label3"="value3"}` string in each lua vm.

`for i in $(seq 1 5); do TEST=$i resty --shdict 'prometheus 10M' memory.lua; done`
[memory.lua.gz](https://github.com/Kong/kong-plugin-prometheus/files/4386734/memory.lua.gz)

A histogram with current exporter.lua buckets takes 3000 bytes .

So for each service/route consider 3 status code,if we take 300 as an average size for counter, there'll be (3 + 2) * 300 + 3 *3000 = 10500 bytes (3 status code, 2 bandwidth, 3 latencies).

**Comment by fffonion** [comment]
hold to sync with upstream

**Comment by fffonion** [comment]
this can be reviewed now

**Comment by hbagdi** [comment]
```diff
diff -Naur kong/plugins/prometheus/prometheus.lua ~/play/nginx-lua-prometheus/prometheus.lua
--- kong/plugins/prometheus/prometheus.lua      2020-04-20 15:14:35.428147739 -0700
+++ /home/hbagdi/play/nginx-lua-prometheus/prometheus.lua       2020-04-20 15:11:57.085112212 -0700
@@ -678,4 +678,4 @@
     "Error while setting '", key, "' to '", value, "': '", err, "'")
 end
 
-return Prometheus
\ No newline at end of file
+return Prometheus

```

@fffonion There is a small diff between the vendored `prometheus.lua` file and the one in the master branch on the upstream. Can you fix that?
This is a huge improvement and this PR makes me very happy! Not just from the performance point of view but also seeing that we are completely aligned with the upstream module.

File: .travis.yml [hunk 1]
```diff
@@ -26,36 +23,37 @@ services:
 
 env:
   global:
+    - PLUGIN_NAME=prometheus
     - LUAROCKS=3.1.3
-    - OPENSSL=1.1.1c
+    - OPENSSL=1.1.1d
+    - KONG_REPOSITORY=kong
+    - KONG_LATEST=master
```
Comment (hbagdi) [inline .travis.yml]: Any reason to not run this against the `next` branch of `kong/kong`?

File: spec/02-access_spec.lua [hunk 2]
```diff
@@ -90,7 +90,7 @@ describe("Plugin: prometheus (access)", function()
         path    = "/metrics",
       })
       local body = assert.res_status(200, res)
-      return body:find('kong_http_status{code="200",service="mock-service",route="http-route"} 1', nil, true)
```
Comment (hbagdi) [inline spec/02-access_spec.lua:93]: curious, why was this changed?

File: kong/plugins/prometheus/exporter.lua [hunk 3]
```diff
@@ -52,29 +51,26 @@ local function init()
...
-  counter_instance, err = counter.new("prometheus_metrics", 1)
-  if err then
-    error(err)
-  end
-  prometheus:set_resty_counter(counter_instance)
+  prometheus:init_worker()
 end
 
+
+local table_3 = {0, 0, 0}
```
Comment (hbagdi) [inline kong/plugins/prometheus/exporter.lua]: `labels_table` might be a better name?

File: spec/02-access_spec.lua (hunk 2 above)
Comment (hbagdi) [reply spec/02-access_spec.lua:93]: I see why now. You swapped the order of labels. Why was that necessary?
Does changing table_3[3] over table_3[1] have a performance difference in Lua/LuaJIT?

Comment (fffonion) [reply spec/02-access_spec.lua:93]: this is to save some memory usage, since in the prometheus library we create a new table for each diverged label, since we expect more than one status code per metrics, so putting the status code at the end will save us some tables being created. I'll put this in the comment

**Review by hbagdi — APPROVED** [review]

**Comment by hbagdi** [comment]
This is a big deal patch. Thanks, @fffonion.
